sdks/java/core/src/main/java/org/apache/beam/sdk/util/RowStringInterpolator.java

```diff
@@ -0,0 +1,181 @@
+ * placeholders (i.e. no curly braces), it will simply return that String, untouched.
+ *
+ * <p>Nested fields can be specified using dot-notation (e.g. {@code "top.middle.nested"}).
+ *
+ * <p>Configure a {@link RowStringInterpolator} like so:
+ *
+ * <pre>{@code
+ * String template = "unified {foo} and {bar.baz}!";
+ * Row inputRow = {foo: "batch", bar: {baz: "streaming"}, ...};
+ *
+ * RowStringInterpolator interpolator = new RowStringInterpolator(template, beamSchema);
+ * String output = interpolator.interpolate(inputRow);
+ * // output --> "unified batch and streaming!"
+ * }</pre>
+ *
+ * <p>Additionally, {@link #interpolate(Row, BoundedWindow, PaneInfo, Instant)} can be used in
+ * streaming scenarios to substitute windowing metadata into the template String. To make use of
+ * this, use the relevant placeholder:
+ *
+ * <ul>
+ *   <li>$WINDOW: the window's string representation
+ *   <li>$PANE_INDEX: the pane's index
+ *   <li>$YYYY: the element timestamp's year
+ *   <li>$MM: the element timestamp's month
+ *   <li>$DD: the element timestamp's day
+ * </ul>
+ *
+ * <p>For example, your Sting template can look like:
+ *
+ * <pre>{@code "unified {foo} and {bar} since {$YYYY}-{$MM}!"}</pre>
+ */
+public class RowStringInterpolator implements Serializable {
+  private final String template;
+  private final Set<String> fieldsToReplace;
+  public static final String WINDOW = "$WINDOW";
+  public static final String PANE_INDEX = "$PANE_INDEX";
+  public static final String YYYY = "$YYYY";
+  public static final String MM = "$MM";
+  public static final String DD = "$DD";
+  private static final Set<String> WINDOWING_METADATA =
+      Sets.newHashSet(WINDOW, PANE_INDEX, YYYY, MM, DD);
+
+  public RowStringInterpolator(String template, Schema rowSchema) {
+    this.template = template;
+
+    Matcher m = Pattern.compile("\\{(.+?)}").matcher(template);
+    fieldsToReplace = new HashSet<>();
+    while (m.find()) {
+      fieldsToReplace.add(StringUtils.strip(m.group(), "{}"));
+    }
+
+    List<String> rowFields =
+        fieldsToReplace.stream()
+            .filter(f -> !WINDOWING_METADATA.contains(f))
+            .collect(Collectors.toList());
+
+    RowFilter.validateSchemaContainsFields(rowSchema, rowFields, "string interpolation");
+  }
+
+  /** Performs string interpolation on the template using values from the input {@link Row}. */
+  public String interpolate(Row row) {
+    String interpolated = this.template;
+    for (String field : fieldsToReplace) {
+      List<String> levels = Splitter.on(".").splitToList(field);
+
+      Object val = MoreObjects.firstNonNull(getValue(row, levels, 0), "");
+
+      interpolated = interpolated.replace("{" + field + "}", String.valueOf(val));
+    }
+    return interpolated;
+  }
+
+  /** Like {@link #interpolate(Row)} but also potentially include windowing information. */
+  public String interpolate(Row row, BoundedWindow window, PaneInfo paneInfo, Instant timestamp) {
```

Comment:
Non blocking but I was concerned about the possibility for future new sources of interpolation. For example, what if someone wants to add a new Thing but include the existing sources of interpolate, row, window, etc? Then in another future, someone else wants to add a new Thing2. This may require some thought. However, what I was envisioning was something like this:

```
interface Interpolator {
    String interpolate();
}
// instantiated with a helper method maybe
class RowInterpolator {
   String interpolate() { }
}
class RowStringInterpolator 
    public String interpolate(Interpolator ...interpolators) { ... }
}
String result = interpolator.interpolate(rowInterpolator, windowInterpolator, paneInfoInterpolator, timestampInterpolator);
```

Reply:
Yeah I could see this being something like

```java
interface Interpolator<T> {
  String interpolate(T);
  ...
}

RowStringInterpolator implements Interpolator<Row> {}
```

With that said though, I'd prefer we leave it off until it is warranted. Most of this work is just for cross-language and portability, where Beam Row is used as a standard. I haven't come across a case where we would benefit with String interpolation using another type, so modularizing this may not actually make a difference.